libraries/apollo-api/src/jsMain/kotlin/com/apollographql/apollo3/api/json/DynamicJsJsonReader.kt

```diff
@@ -0,0 +1,64 @@
+package com.apollographql.apollo3.api.json
+
+
+@Suppress("NOTHING_TO_INLINE")
+private inline fun objectKeys(json: dynamic) = js("Object").keys(json).unsafeCast<Array<String>>()
+
+
+class DynamicMapEntry<K, V>(
+    override val key: K,
+    override val value: V
+) : Map.Entry<K, V>
+
+/**
+ * A [JsonReader] that reads data from a regular JS `dynamic` object.
+ *
+ * Dynamic values should be any of:
+ * - String
+ * - Int
+ * - Double
+ * - Long
+ * - JsonNumber
+ * - null
+ * - `dynamic` where values are any of these values recursively
+ * - Array<Any?> where values are any of these values recursively
+ *
+ * Anything else is undefined
+ *
+ * TODO Note: We can probably safely remove the non-JS primitives (ie Long, Double, JsonNumber)
+ *
+ */
+class DynamicJsJsonReader constructor(
+    root: dynamic,
+    pathRoot: List<Any> = emptyList(),
+) : MapishJsonReader<dynamic>(
+    root = root,
+    pathRoot = pathRoot
+) {
+  override fun anyToToken(any: Any?) = when (any) {
+    null -> JsonReader.Token.NULL
+    is Array<*> -> JsonReader.Token.BEGIN_ARRAY
+    is Int -> JsonReader.Token.NUMBER
+    is Long -> JsonReader.Token.LONG
+    is Double -> JsonReader.Token.NUMBER
+    is JsonNumber -> JsonReader.Token.NUMBER
+    is String -> JsonReader.Token.STRING
+    is Boolean -> JsonReader.Token.BOOLEAN
+    else -> {
+      if (objectKeys(any).isNotEmpty()) {
+        JsonReader.Token.BEGIN_OBJECT
+      } else {
+        JsonReader.Token.ANY
+      }
+    }
+  }
+
+  override fun getArrayIterator(container: Any): Iterator<Any?> {
+    @Suppress("UNCHECKED_CAST")
+    return (container as Array<Any?>).iterator()
+  }
+
+  override fun getMapIterator(container: dynamic): Iterator<Map.Entry<String, Any?>> {
+    return objectKeys(container).map { key -> DynamicMapEntry(key, container[key]) }.iterator()
```

Comment:
Pretty sure we can optimize by not going through `Iterator` and saving some `DynamicMapEntry` but that certainly means duplicating some code.

Reply:
Ah, yes indeed. From looking at the profiles it doesn't seem like we're spending a ton of time there. Most of the time is spent in the data class construction. This change speeds up the parsing step by ~15x, so I'm not sure we need to squeeze more out of it.

Happy to make the change though if you think it is worth the tradeoff.